### DISCUSSION TITLE
Performance Issues - optimization alpha or python problems?

### DISCUSSION THREAD

**Discussion by KeebCat** [TagStudioDev/TagStudio#309] [General]
As has been stated in the YouTube video announcing this project, there used to be performance issues with the Python GUI library that was previously used, and these problems were inherent to it. While this program runs fine normally, at larger library sizes it does behave more sluggishly. Is this just a side effect of this project being in an alpha state or could it be a problem with Python not being the best language for applications which require high performance?

Or in other words: Have you considered performance with very large image libraries? Is Python the right language?

As I am not a contributor to this project I do not presume that I know better than anyone here, I'm just wondering if this has been thought about? I'd like to raise this question early, because in case Python does end up being considered a bottleneck, it'd probably be best to switch early, and if not and this is just a side effect of being in an alpha state, at least the consideration is out of the way.

As a side note, thank you to all contributors for the work on this project. It shows a lot of promise and I wish Tag Studio the best <3

**Comment by CyanVoxel** [comment]
Thank you for checking out the project! The vast majority of the current UI bottlenecking comes down to the lack of thumbnail caching, something that will be added in an upcoming version. Reading the contents of large file directories upon library creation and refreshing is also less due to Python but more of a hard limit of the OS/file system. With that being said, I'm also looking into leveraging MFT tables on NFTS to help dramatically speed this up on supported systems.

All in all though most of the current sluggish areas are due to factors other than Python, much to even my surprise. I didn't initially intend to stick with Python this far into development but so far it hasn't been a real bottleneck. And since switching to Qt as the frontend the QUI has been *much* more responsive as (I believe) it's using C bindings compared to TKinter which as entirely Python.

**Reply by KeebCat** [reply]
Good to know, thank you for the info!
